.github/workflows/set-metadata.yml

```diff
@@ -0,0 +1,385 @@
+name: Reusable WF - Set Build Metadata
+
+on:
+  workflow_call:
+    inputs:
+      build-type:
+        required: false
+        type: string
+      payload:
+        required: false
+        type: string
+      optional-release-tag:
+        required: false
+        type: string
+      amalgam-build:
+        required: false
+        type: string
+      howso-engine-build:
+        required: false
+        type: string
+      amalgam-lang-py-build:
+        required: false
+        type: string
+      howso-engine-py-build:
+        required: false
+        type: string
+      howso-synthesizer-py-build:
+        required: false
+        type: string
+      howso-validator-py-build:
+        required: false
+        type: string
+      howso-validator-enterprise-py-build:
+        required: false
+        type: string
+    outputs:
+      upstream-details:
+        value: ${{ jobs.set-metadata.outputs.upstream-details }}
+      version:
+        value: ${{ jobs.set-version.outputs.version }}
+
+jobs:
+
+  construct-payload:
+    runs-on: ubuntu-latest
+    outputs:
+      payload: ${{ steps.construct-payload.outputs.payload }}
+    steps:
+
+    - name: Construct payload
+      id: construct-payload
+      run: |
+        if [[ -n "${{ inputs.payload }}" ]]; then
+          echo "Found preconstructed JSON payload"
+          echo "payload=${{ inputs.payload }}" >> $GITHUB_OUTPUT
+          exit 0
+        fi
+
+        echo "Constructing JSON payload from raw inputs"
+        # Map inputs to repo names in a JSON object
+        builds_json=$(cat <<EOF
+        {
+          "amalgam": "${{ inputs.amalgam-build }}",
+          "amalgam-lang-py": "${{ inputs.amalgam-lang-py-build }}",
+          "howso-engine": "${{ inputs.howso-engine-build }}",
+          "howso-engine-py": "${{ inputs.howso-engine-py-build }}",
+          "howso-synthesizer-py": "${{ inputs.howso-synthesizer-py-build }}",
+          "howso-validator-py": "${{ inputs.howso-validator-py-build }}",
+          "howso-validator-enterprise-py": "${{ inputs.howso-validator-enterprise-py-build }}",
+        } 
+        EOF
+        )
+        # Initialize an empty final payload
+        payload="{}"
+
+        # If the provided build is a URL, extract the run ID. Else, if the build exists, add it to the payload JSON.
+        while read -r repo; do
+          build=$(echo "$builds_json" | jq --arg repo "$repo" '".$repo"')
+          echo "Discovered build '$build' for repo $repo"
+          if [[ "$build" =~ runs/([0-9]+) ]]; then
+            payload=$(echo "$payload" | jq --arg run_id "${BASH_REMATCH[1]}" --arg repo "$repo" '. + {"$repo": $run_id}')
+          elif [[ -n "$build" ]]; then
+            payload=$(echo "$payload" | jq --arg build "$build" --arg repo "$repo" '. + {"$repo": $build}')
+          fi
```

Comment:
Since this is user input, probably good to have an 'else' that errors out with a helpful message.

Reply:
So sometimes $build could be empty and it could be valid -- for example, `howso-validator-py-build` would be an empty string for all runs of `howso-engine-py`. We just need to check for all of them in case it is a higher-level project. Unless I'm misunderstanding my Bash, the `elif` there will trigger as long as it's not empty.

Reply:
Ah, didn't catch that $build being empty is legit case, sounds good.